keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
+//    maintenance in progress.
+Complete NodeMaintenanceStage = "Complete"
+)
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // NodeSelector selects nodes for this node maintenance.
+    // +required
+    NodeSelector *v1.NodeSelector
+
+    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
+    //
+    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
+    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
+    //
+    // The default value is Idle.
+    Stage NodeMaintenanceStage
+
+    // DrainPlan is executed from the first entry to the last entry during the Drain stage.
+    // DrainPlanEntry podType fields should be in the following order:
+    // nil -> DaemonSet -> Static
+    // DrainPlanEntry priority fields should be in ascending order for each podType.
+    // If the priority and podType are the same, concrete selectors are executed first.
+    //
+    // The following entries are injected into the drainPlan on the NodeMaintenance admission:
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Default"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Default"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "DaemonSet"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "DaemonSet"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Static"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Static"
+    //
+    // Duplicate entries are not allowed.
+    // This field is immutable.
+    DrainPlan []DrainPlanEntry
+
+    // Reason for the maintenance.
+    Reason string
+}
+
+const (
+// Default selects all pods except DaemonSet and Static pods.	
+Default PodType = "Default"
+// DaemonSet selects DaemonSet pods.
+DaemonSet PodType = "DaemonSet"
+// Static selects static pods.
+Static PodType = "Static"
+)
+
+type DrainPlanEntry struct {
```

Comment:
I have question/doubt about this API.

1) If I got it right, in order to define a rule that targets a specific set of Pods, it is required that those Pods have an identifying PodPriority. Is this correct? 

If true, what if a user needs to define a drain rule for an application where it can define a pod priority (the users not can change freely pods in applications, e.g. if the Pods are managed by some operator)?

Also, is it correct to introduce a dependency between PodPriority and drain? what if users wants to define a drain rule for an application, but they don't want that this application triggers preemption for other apps when scheduled? Are we mixing up two orthogonal problems?

2) It seems also that each rule Must be specifically defined for a single PodType, Is this correct?

if true, I'm worried that we are surfacing is API that some of the internals of the system, the fact that different K8s components takes care of different type pods.

As a user, I could find really useful to define a specific drain rule for my application, but when I try to image such rule I think at something like "I want to evict Portworx after evicting all the other pods". 

I would expect it to be a responsibility of the system to figure out what kind of Pods my application has, and do the right thing.

But if instead we expect users to dig into applications, usually designed by other teams, might be not even in my company, figure out what kind of pods the application has, and the map this to all the intricacies of K8s components and node drain, this could really impact the usability of this feature.

Reply:
Good feedback. I'd be very happy to see an initial alpha API (either in-tree or a CRD) where we can practice triggering declarative draining, and encourage early adopters to give us their thoughts whilst the new API is still plastic (mouldable, fixable).

Reply:
To me, declarative node maintenance is / should be much more about:
- I intend to empty this node / this rack / everything with this storage atttached to it
- here's when it will happen, and maybe here's a set of taints that are likely to get implied
  - might be good to know in case your app intends to tolerate these
- controllers are welcome to handle this gracefully before k-c-m (I assume) starts telling the kubelet to delete Pods

and out of all of that, the most key detail to capture is that it is an intent to maintain, rather than a verb masquerading as a noun (in other words, we aren't making a NodeDrainRequest or a PodRemovalSchedule).

Reply:
Makes sense.
Might be the best way to bridge the two worlds is to think to a two step process/two CRDs:
- Allow application developers to declare Drain rules for their applications, like they already do for PDBs
- Allow cluster operators to declare node maintenance intent, which will then collect and use Drain rules if necessary

Also, FYI, in Cluster API we are thinking about something to make our machine drain process more flexible, and due to a few constraints, we need it in the short term; however, I can easily image how Cluster API users might benefit from something similar to what we are discussing in this thread (when it becomes available).

Reply:
I think the point of the DrainPlan is not to define rules on how to gracefully terminate an application. This is an admin level tool that is more coarse and basically just looks at different classes of applications (e.g. user vs system). And can differentiate between different control-plane components.

The user does not have to tinker with the priorities to achieve the desired result. They also have no guarantee that the NodeMaintenance will be the same in every cluster (e.g. all of the user app pods can fall into the same termination bucket).

To solve this, users can simply run all of their pods at e.g. priority 0 an utilize the [Evacuation API](https://github.com/kubernetes/enhancements/pull/4565) logic. Users can add an evacuator to their app pods and downscale the app according to their needs. For example, the evacuator can keep pod that provides a service alive until a pod that depends on that service terminates.

NodeMaintenance and Evacuation are intended to be extensible and for users to be able to depend on the presence of these APIs in the cluster (similar to eviction + PDB). This is the main reason why it is hard to introduce this as a CRD.

Reply:
I agree declaring node maintenance is an admin operation.
But if node maintenance will only be about corse configuration (user vs system or deamon set vs static pods vs everything else), it will fail to address the friction that exists between cluster admin and application teams when a node must be drained.

WRT to the evacuation API, my personal take is that this is more a solution for advanced use cases like operators. The majority of folks need some more generic, and DrainRules have the potential to fill this gap (but in order to do so, they must be decoupled by NodeMaintenance)

Reply:
Yes there is a friction between the cluster admin and applicaiton teams. But it is quite a difficult task to offer a strictly declarative API to support all possible graceful draining scenarios. There could be a lot of dependencies and edge cases. I am not convinced we want to support that.

The Evacuation API allows you to do anything, but I agree that it is a quite advanced/handful for a normal user. That is why we will try to offer clever draining behaviors for some workloads (e.g. Deployment, HPA, configurable StatefulSet, etc.). And also, the Evacuation API can be used as a building block. So other components in the ecosystem can implement DrainRules and offer it as an Operator/CRD that normal users can use for a subset of applications.

Reply:
Thank you for the explanation, I appreciate your patience in following my questions.

However, if what we need to solve this problem for both cluster admin and application teams is to eventually build yet another layer on top of the evacuation API, frankly speaking IMO the complexity is just too much (it is hard to figure out how m:n rules works in this proposal, let's image another layer of rules on top).

If I combine this with proposed breaking changes like draining daemonst sets and static pods, and a few other thread of discussion, it is hard to figure out how would look like the final state with a cohesive solution that helps both cluster admin and application teams, and also when this will be available.

Considering this, I'm starting to lean towards the opinion already expressed by other contributors providing feedback on this proposal that NodeMaintenance in the current form should be incubated as an external CRD, and then move to k/k when it is mature. But this depends also by how other thread are addressed.

In the meantime, I kindly ask you to add the idea discussed in this thread as alternatives in the proposal:
- decoupling drain rules definition from the node maintenance initiation (the split is what enables multiple personas to play part in this use case)
- make drain rule more application friendly (also avoiding to surface K8s internals in the user facing API)

cc @vincepri @neolit123 @sbueringer @chrischdi given that this popped up in a few SIG Cluster lifecycle related discussions

Reply:
> If what we need to solve this problem for both cluster admin and application teams is to eventually build yet another layer on top of the evacuation API, frankly speaking IMO the complexity is just too much (it is hard to figure out how m:n rules works in this proposal, let's image another layer of rules on top).

Another layer (whatever it is) above the Evacuation API is necessary as the Evacuation API servers as a base and implements graceful drain which is mostly not possible today.

> If I combine this with proposed breaking changes like draining daemonst sets and static pods,

Can you please let me know which part do you consider a breaking change?

> decoupling drain rules definition from the node maintenance initiation (the split is what enables multiple personas to play part in the solution of this problem)

I think it makes sense to decouple the rules. We probably still want to keep the admission logic. We could also introduce a `Default` DrainPlan that would be selected if none was supplied to the NodeMaintenance as I would expect it might be hard to select the correct one for some drainers.

> make drain rule more application friendly (also avoiding to surface K8s internals in the user facing API)

Application user facing API could be a nice followup if the community finds it useful.